TimeLoopLib/ExternalDynamicTimeSteppingIndividual.py

```diff
@@ -0,0 +1,102 @@
+    # n_step_bg = 1, n_step_gd = 1, then above-ground competition is updated
+    # every 2nd step while the other concepts are updated in each time step.
+    def setSteps(self, n_step_ag, n_step_bg, n_step_gd):
+        self.n_step_ag = n_step_ag
+        self.n_step_bg = n_step_bg
+        self.n_step_gd = n_step_gd
+
+    ## This progresses one time step, by updating tree population and above-
+    # and below-ground resources. That single function calls are similar to
+    # the basic MANGA time stepper (TreeDynamicTimeStepping) with the
+    # difference that not all concepts have to be called with the same
+    # frequence.
+    def step(self, t_end):
+        t_start = self.t_step_begin
+        self.t_step_begin = t_end
+
+        update_ag = True if self.step_counter % self.n_step_ag == 0 else False
+        update_bg = True if self.step_counter % self.n_step_bg == 0 else False
+        update_gd = True if self.step_counter % self.n_step_gd == 0 else False
+
+        tree_groups = self.population.getTreeGroups()
+        self.tree_output.writeOutput(tree_groups, t_start)
+
+        if update_ag:
+            self.aboveground_competition.prepareNextTimeStep(t_start, t_end)
+        if update_bg:
+            self.belowground_competition.prepareNextTimeStep(t_start, t_end)
+        if update_gd:
+            self.death_and_growth_concept.prepareNextTimeStep(t_start, t_end)
+
+        for group_name, tree_group in tree_groups.items():
+            for tree in tree_group.getTrees():
+                if update_ag:
+                    self.aboveground_competition.addTree(tree)
+                if update_bg:
+                    self.belowground_competition.addTree(tree)
+
+        if update_ag:
+            self.aboveground_competition.calculateAbovegroundResources()
+            self.aboveground_resources = (
+                self.aboveground_competition.getAbovegroundResources())
+        if update_bg:
+            self.belowground_competition.calculateBelowgroundResources()
+            self.belowground_resources = (
+                self.belowground_competition.getBelowgroundResources())
+        if update_gd:
+            j = 0
+            for group_name, tree_group in tree_groups.items():
+                kill_indices = []
+                for tree, i in zip(tree_group.getTrees(),
+                                   range(tree_group.getNumberOfTrees())):
+                    # if a new tree is recruited in the current time step and
+                    # the respective resource was not updated, set survival of
+                    # the new tree to 1
+                    try:
+                        ag = self.aboveground_resources[j]
+                        bg = self.belowground_resources[j]
+                        self.death_and_growth_concept.progressTree(
+                            tree, ag, bg)
+                    except IndexError:
+                        tree.setSurvival(1)
```

Comment:
ok. what could go wrong here?
maybe you could add something like
```
if tree_propertie_proves_it_is_new: tree.setSturvival(1)
else: raise IndexError 
```
?

Reply:
Do you mean that a tree that just died could survive?

Reply:
no. I am afraid that we incorporate a bug here which might be hard to find as no error will be raised. So we need to be as sure as possible that the tree is a new one.

Reply:
And do you have an idea how to check this? I can only think of the index...

Reply:
Maybe sth like this?

```suggestion
                    if tree.getId in trees_living_more_than_1_dt:
                        ag = self.aboveground_resources[j]
                        bg = self.belowground_resources[j]
                        self.death_and_growth_concept.progressTree(
                            tree, ag, bg)
                    else:
                        trees_living_more_than_1_dt.append(tree.getId())
                        tree.setSurvival(1)
```